### PR TITLE
Add support for padding vertex attributes

### PR DESCRIPTION
**Pull request by ds84182**
This adds support for padding attributes, as defined on [3dbrew](https://www.3dbrew.org/wiki/GPU/Internal_Registers#GPUREG_ATTRIBBUFFERi_CONFIG1). I created a small example (which runs on real hardware) [here](https://github.com/ds84182/3ds-gpu-stuff/tree/master/attribute_padding).

### CODE REVIEW

**Comment by neobrain** [comment]
Hi, this is neobot, using neobrain's account. I'm keeping an archive of versions of this PR:

2016-02-06T00:13:49Z: https://github.com/citra-emu/citra/compare/69e2453649b30962ac68ef54ffbf30ab4f1a9842...ds84182:dc8d5617099f6e26ded5bdec47dbe7c460e2ba34

File: src/video_core/command_processor.cpp [hunk 1]
```diff
@@ -160,12 +160,17 @@ static void WritePicaReg(u32 id, u32 value, u32 mask) {
...
                     u32 attribute_index = loader_config.GetComponent(component);
-                    vertex_attribute_sources[attribute_index] = load_address;
-                    vertex_attribute_strides[attribute_index] = static_cast<u32>(loader_config.byte_count);
-                    vertex_attribute_formats[attribute_index] = attribute_config.GetFormat(attribute_index);
-                    vertex_attribute_elements[attribute_index] = attribute_config.GetNumElements(attribute_index);
-                    vertex_attribute_element_size[attribute_index] = attribute_config.GetElementSizeInBytes(attribute_index);
-                    load_address += attribute_config.GetStride(attribute_index);
+                    if (attribute_index >= 12) {
+                        // Padding attribute
+                        load_address += (attribute_index-11)*4;
```
Comment (lioncash) [inline src/video_core/command_processor.cpp]: space before and after `-` and `*`

**Comment by yuriks** [comment]
@JohnGodgames: Does this fix #974?

**Comment by JohnGodgames** [comment]
nope...

**Comment by neobrain** [comment]
Hi, this is neobot, using neobrain's account. I'm keeping an archive of versions of this PR:

2016-02-06T15:03:13Z: https://github.com/citra-emu/citra/compare/f1d1049c4fdf51fd382f9afb6b5cd5f104e71b0e...ds84182:04ab188d64ed5c667c780197d7a42a8465f3e02a

File: src/video_core/command_processor.cpp [hunk 2]
```diff
@@ -160,12 +160,17 @@ static void WritePicaReg(u32 id, u32 value, u32 mask) {
...
                     u32 attribute_index = loader_config.GetComponent(component);
-                    vertex_attribute_sources[attribute_index] = load_address;
-                    vertex_attribute_strides[attribute_index] = static_cast<u32>(loader_config.byte_count);
-                    vertex_attribute_formats[attribute_index] = attribute_config.GetFormat(attribute_index);
-                    vertex_attribute_elements[attribute_index] = attribute_config.GetNumElements(attribute_index);
-                    vertex_attribute_element_size[attribute_index] = attribute_config.GetElementSizeInBytes(attribute_index);
-                    load_address += attribute_config.GetStride(attribute_index);
+                    if (attribute_index >= 12) {
+                        // Padding attribute
+                        load_address += (attribute_index - 11) * 4;
```
Comment (bunnei) [inline src/video_core/command_processor.cpp]: Can you change this to be a little more self-documenting? E.g. something like:
<pre>
switch (attribute_index) {
case 0: // Valid vertex attribute indices are 0-11
case 1:
case 2:
case 3:
case 4:
case 5:
case 6:
case 7:
case 8:
case 9:
case 10:
case 11:
    vertex_attribute_sources[attribute_index] = load_address;
    vertex_attribute_strides[attribute_index] = static_cast<u32>(loader_config.byte_count);
    vertex_attribute_formats[attribute_index] = attribute_config.GetFormat(attribute_index);
    vertex_attribute_elements[attribute_index] = attribute_config.GetNumElements(attribute_index);
    vertex_attribute_element_size[attribute_index] = attribute_config.GetElementSizeInBytes(attribute_index);
    load_address += attribute_config.GetStride(attribute_index);
	break;
	
case 12: // 4-byte padding
case 13: // 8-byte padding
case 14: // 12-byte padding
case 15: // 16-byte padding
	load_address += (attribute_index - 11) * 4;
	break;
	
default:
	LOG_CRITICAL(HW_GPU, "Unknown vertex attribute index %d", attribute_index);
    UNREACHABLE();
	break;
}
</pre>
A little verbose, but this makes the intended functionality much more clear :)

**Comment by bunnei** [comment]
LGTM after the above comment is addressed

File: src/video_core/command_processor.cpp [hunk 3]
```diff
@@ -160,12 +160,17 @@ static void WritePicaReg(u32 id, u32 value, u32 mask) {
                     if (component >= 12)
```
Comment (mailwl) [inline src/video_core/command_processor.cpp]: i think this check not need anymore

File: src/video_core/command_processor.cpp (hunk 2 above)
Comment (yuriks) [inline src/video_core/command_processor.cpp]: Eh... I'm not too amazed by this. It explodes the code with verbosity and redundancy, and, to me, isn't even an improvement. I'd just expand on the comment at the top of the if to note "// Attribute ids 12, 13, 14 and 15 signify 4, 8, 12 and 16-byte paddings, respectively"

**Comment by neobrain** [comment]
Hi, this is neobot, using neobrain's account. I'm keeping an archive of versions of this PR:

2016-02-13T21:47:08Z: https://github.com/citra-emu/citra/compare/f1d1049c4fdf51fd382f9afb6b5cd5f104e71b0e...ds84182:630a24cd89462657eea0c7330d6f329b5f928ca7

File: src/video_core/command_processor.cpp (hunk 3 above)
Comment (bunnei) [inline src/video_core/command_processor.cpp]: Ditto... I noticed you updated this PR without addressing this comment?

File: src/video_core/command_processor.cpp [hunk 4]
```diff
@@ -160,12 +160,17 @@ static void WritePicaReg(u32 id, u32 value, u32 mask) {
...
                         LOG_ERROR(HW_GPU, "Overflow in the vertex attribute loader %u trying to load component %u", loader, component);
                     u32 attribute_index = loader_config.GetComponent(component);
-                    vertex_attribute_sources[attribute_index] = load_address;
-                    vertex_attribute_strides[attribute_index] = static_cast<u32>(loader_config.byte_count);
-                    vertex_attribute_formats[attribute_index] = attribute_config.GetFormat(attribute_index);
-                    vertex_attribute_elements[attribute_index] = attribute_config.GetNumElements(attribute_index);
-                    vertex_attribute_element_size[attribute_index] = attribute_config.GetElementSizeInBytes(attribute_index);
-                    load_address += attribute_config.GetStride(attribute_index);
+                    if (attribute_index >= 12) {
```
Comment (bunnei) [inline src/video_core/command_processor.cpp]: I think the solution I provided was more explicit, but meh, the comment below at least documents what going on here. I think this should still be changes to 3 if conditions, though:
</pre>
if >= 0 and < 12: Normal vertex attribute
elsif >= 12 and < 16: Padding
else: Critical error / Unreachable
</pre>

File: src/video_core/command_processor.cpp (hunk 3 above)
Comment (yuriks) [inline src/video_core/command_processor.cpp]: This check is unrelated to the target attribute index, they just both happen to bound at 12.

File: src/video_core/command_processor.cpp [hunk 5]
```diff
@@ -160,12 +160,17 @@ static void WritePicaReg(u32 id, u32 value, u32 mask) {
                     if (component >= 12)
                         LOG_ERROR(HW_GPU, "Overflow in the vertex attribute loader %u trying to load component %u", loader, component);
```
Comment (yuriks) [inline src/video_core/command_processor.cpp]: Just realized. This should `continue` when this message is printed so that the remaining code doesn't make out of bounds array accesses.

Comment (yuriks) [inline src/video_core/command_processor.cpp]: (This is not introduced by your changes, but might as well fix it here. I'd encourage making it a separate commit in the same PR.)

File: src/video_core/command_processor.cpp (hunk 4 above)
Comment (yuriks) [inline src/video_core/command_processor.cpp]: I agree about treating the unreachable case, though most of the tests there are redundant (and >= 0 will even warn, for example). I suggest:
```
if (attribute_index < 12) {
  // Load attribute
} else if (attribute_index < 16) {
  // Padding
} else {
  UNREACHABLE(); // This is truly unreachable due to the number of bits for each component, so I don't think a log is needed.
}
```

**Comment by neobrain** [comment]
Hi, this is neobot, using neobrain's account. I'm keeping an archive of versions of this PR:

2016-02-21T00:03:35Z: https://github.com/citra-emu/citra/compare/0d086616d1af239b8e41b246bea3c1ef85fc907f...ds84182:ed8072b48bf6eacb0eea161f87746086401b5c08
